<div style="text-align: center; background-color: #FF0000; font-family: 'Georgia', serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
    PROGRAMMING FOR DATA SCIENCE
  <div style="text-align: center; background-color: #FF0000; font-family: 'Georgia', serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
    @ FIT-HCMUS, VNU-HCM 
  </div>
  <div style="text-align: center; background-color: #FF0000; font-family: 'Georgia', serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
    FINAL PROJECT 
  </div>
</div>

<div style="text-align: center; background-color: #ffc8a0; font-family: Verdana, sans-serif; color: white; padding: 20px; font-size: 30px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
    DATA EXPLORATION & DATA PREPROCESSING
</div>

## Đọc dữ liệu vào dataframe
Dữ liệu: [Red Wine Quality](https://www.kaggle.com/api/v1/datasets/download/uciml/red-wine-quality-cortez-et-al-2009).
Theo mô tả về bộ dữ liệu [Red Wine Quality Description](https://www.kaggle.com/datasets/uciml/red-wine-quality-cortez-et-al-2009/data):
> Mục tiêu: xây dựng mô hình dự đoán chất lượng cảm quan (sensory quality) của rượu vang dựa trên các thông số hóa lý.
> 
> Bộ dữ liệu này bao gồm:
> * Số mẫu (records): 1,599
> * Số thuộc tính (features): 12 (bao gồm các thông số hóa lý và nhãn chất lượng)
> 
> Bộ dữ liệu bao gồm 11 thuộc tính đầu vào, cụ thể như sau:  
> * Fixed acidity: Độ axit cố định
> * Volatile acidity: Độ axit dễ bay hơi
> * Citric acid: Hàm lượng axit citric
> * Residual sugar: Hàm lượng đường còn lại
> * Chlorides: Hàm lượng chloride
> * Free sulfur dioxide: Lượng lưu huỳnh dioxide tự do
> * Total sulfur dioxide: Tổng lượng lưu huỳnh dioxide
> * Density: Khối lượng riêng
> * pH: Độ pH
> * Sulphates: Hàm lượng sulphate
> * Alcohol: Nồng độ cồn

In [1]:
import numpy as np
import pandas as pd

In [2]:
# Đọc dữ liệu từ tệp winequality-red.csv vào khung dữ liệu red_wine_quality_df
red_wine_quality_df = pd.read_csv('./winequality-red.csv')

# In một số mẫu
red_wine_quality_df.head(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

## Có bao nhiêu cột và dòng?

In [3]:
# Số hàng và cột dữ liệu
print("Rows and columns:", red_wine_quality_df.shape)

n_rows = red_wine_quality_df.shape[0]
n_cols = red_wine_quality_df.shape[1]

print("Rows:", n_rows)
print("Columns:", n_cols)

Rows and columns: (1599, 12)
Rows: 1599
Columns: 12


## Dòng
### Ý nghĩa của từng dòng?
Mỗi dòng sử dụng mẫu rượu vang đỏ. Đầu vào bao gồm các thử nghiệm khách quan (ví dụ: giá trị PH) và đầu ra dựa trên dữ liệu cảm quan (trung bình của ít nhất 3 đánh giá do các chuyên gia về rượu thực hiện). Mỗi chuyên gia đánh giá chất lượng rượu từ 0 (rất kém) đến 10 (rất tuyệt vời).  
### Có dòng bị trùng lặp không?  

In [4]:
# Tìm số dòng trùng lặp
print("Duplicated rows:",red_wine_quality_df.duplicated().sum())

Duplicated rows: 240


Vậy có 240 dòng trùng lặp. Ta cần xóa những dòng trùng lặp đó.

In [5]:
red_wine_quality_df.drop_duplicates(inplace=True)

Kiểm tra lại dữ liệu có còn trùng lặp hay không.

In [6]:
print("Duplicated rows:",red_wine_quality_df.duplicated().sum())

Duplicated rows: 0


Các dòng trùng lặp đã bị xóa. Tiếp theo chúng ta sẽ khám phá các cột.

## Cột
### Ý nghĩa của từng cột?  

| Thuộc tính              | Ý nghĩa                                                       | Đơn vị      | Vai trò                                                | Dạng dữ liệu  | Tính chất    |
|--------------------------|--------------------------------------------------------------|-------------|--------------------------------------------------------|--------------|--------------|
| Fixed acidity           | Độ axit cố định (axit không bay hơi, như axit tartaric)      | g/dm³       | Tạo vị chua và cấu trúc của rượu                      | Numeric        | Continuous   |
| Volatile acidity        | Độ axit dễ bay hơi (chủ yếu là axit axetic)                  | g/dm³       | Ảnh hưởng đến mùi, hàm lượng cao gây mùi giấm         | Numeric        | Continuous   |
| Citric acid             | Hàm lượng axit citric                                        | g/dm³       | Tăng cảm giác tươi mới và cân bằng vị chua            | Numeric        | Continuous   |
| Residual sugar          | Lượng đường còn lại sau quá trình lên men                   | g/dm³       | Quyết định độ ngọt và khả năng bảo quản               | Numeric        | Continuous   |
| Chlorides               | Hàm lượng chloride (muối)                                    | g/dm³       | Ảnh hưởng đến vị mặn, mức cao làm giảm chất lượng     | Numeric        | Continuous   |
| Free sulfur dioxide     | Lưu huỳnh dioxide tự do                                      | mg/dm³      | Chống oxy hóa và vi khuẩn                             | Numeric        | Discrete   |
| Total sulfur dioxide    | Tổng lượng lưu huỳnh dioxide                                 | mg/dm³      | Ảnh hưởng đến độ ổn định và bảo quản rượu             | Numeric        | Discrete   |
| Density                 | Khối lượng riêng của rượu so với nước                        | kg/m³       | Phản ánh lượng đường và cồn                           | Numeric        | Continuous   |
| pH                      | Độ pH của rượu                                               | Không có    | Đo độ axit, ảnh hưởng đến tuổi thọ và hương vị        | Numeric        | Continuous   |
| Sulphates               | Hàm lượng sulphate (muối của axit sulfuric)                 | g/dm³       | Góp phần tạo cảm giác chát nhẹ và giữ ổn định chất lượng | Numeric        | Continuous   |
| Alcohol                 | Nồng độ cồn (ethanol)                                        | % (thể tích)| Quyết định độ đậm đà và cảm giác rượu khi uống        | Numeric        | Continuous   |
| Quality                 | Điểm đánh giá chất lượng cảm quan của rượu vang             | Không có    | Thước đo chất lượng rượu tổng thể                    | Numeric      | Ordinal    |



Dạng dữ liệu của tất cả các cột đều là numeric nên ta không cần xử lí đối với dạng dữ liệu categorical.

### Kiểu dữ liệu của từng cột là gì?

In [7]:
red_wine_quality_df.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

### Có kiểu dữ liệu nào không thích hợp không?
Các kiểu dữ liệu đã thích hợp nhưng giá trị của một số cột có vẻ hơi dài nên ta nên làm tròn đến số thập phân thứ 2.

In [8]:
red_wine_quality_df = red_wine_quality_df.round(2)
red_wine_quality_df.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0             7.4              0.70         0.00             1.9       0.08   
1             7.8              0.88         0.00             2.6       0.10   
2             7.8              0.76         0.04             2.3       0.09   
3            11.2              0.28         0.56             1.9       0.08   
5             7.4              0.66         0.00             1.8       0.08   
6             7.9              0.60         0.06             1.6       0.07   
7             7.3              0.65         0.00             1.2       0.06   
8             7.8              0.58         0.02             2.0       0.07   
9             7.5              0.50         0.36             6.1       0.07   
10            6.7              0.58         0.08             1.8       0.10   

    free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                  11.0                  34.0     1.00  3.51       0.56   
1                  25.0                  67.0     1.00  3.20       0.68   
2                  15.0                  54.0     1.00  3.26       0.65   
3                  17.0                  60.0     1.00  3.16       0.58   
5                  13.0                  40.0     1.00  3.51       0.56   
6                  15.0                  59.0     1.00  3.30       0.46   
7                  15.0                  21.0     0.99  3.39       0.47   
8                   9.0                  18.0     1.00  3.36       0.57   
9                  17.0                 102.0     1.00  3.35       0.80   
10                 15.0                  65.0     1.00  3.28       0.54   

    alcohol  quality  
0       9.4        5  
1       9.8        5  
2       9.8        5  
3       9.8        6  
5       9.4        5  
6       9.4        5  
7      10.0        7  
8       9.5        7  
9      10.5        5  
10      9.2        5

## Đối với mỗi cột có kiểu dữ liệu numeric, các giá trị được phân bổ như thế nào?
Ở đây thì tất cả các cột đều là dữ liệu numeric. Chúng ta cần xem có bao nhiêu giá trị bị thiếu trong các cột.

In [9]:
np.sum(np.isnan(red_wine_quality_df), axis=0)

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

Không có giá trị nào bị thiếu trong tất cả các cột số. Chúng ta sẽ tính giá trị min, Q1(25%), median, Q3(75%) và max của các cột số này.

In [11]:
# Tính toán các giá trị thống kê: min, Q1, median, Q3, max cho tất cả các cột số
summary_stats = red_wine_quality_df.describe(percentiles=[0.25, 0.5, 0.75]).loc[['min', '25%', '50%', '75%', 'max']]

summary_stats

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
min            4.6              0.12         0.00             0.9       0.01   
25%            7.1              0.39         0.09             1.9       0.07   
50%            7.9              0.52         0.26             2.2       0.08   
75%            9.2              0.64         0.43             2.6       0.09   
max           15.9              1.58         1.00            15.5       0.61   

     free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
min                  1.0                   6.0     0.99  2.74       0.33   
25%                  7.0                  22.0     1.00  3.21       0.55   
50%                 14.0                  38.0     1.00  3.31       0.62   
75%                 21.0                  63.0     1.00  3.40       0.73   
max                 72.0                 289.0     1.00  4.01       2.00   

     alcohol  quality  
min      8.4      3.0  
25%      9.5      5.0  
50%     10.2      6.0  
75%     11.1      6.0  
max     14.9      8.0

## <font color=red> Lưu xuống file dữ liệu mới sau tiền xử lí để chuẩn bị cho phần câu hỏi </font>

In [13]:
red_wine_quality_df.to_csv('new-redwine-quality.csv', index = False)